internal/buff/read.go

```diff
@@ -0,0 +1,226 @@
+	data    *soc.Data
+	Err     error
+	Buf     []byte
+	MsgType uint8
+}
+
+// NewReader returns a new Reader.
+func NewReader(toBeDeserialized chan *soc.Data) *Reader {
+	return &Reader{toBeDeserialized: toBeDeserialized}
+}
+
+// SimpleReader creates a new reader that operates on a single []byte.
+func SimpleReader(buf []byte) *Reader {
+	r := &Reader{Buf: buf}
+	(*reflect.SliceHeader)(unsafe.Pointer(&r.Buf)).Cap = len(buf)
+	return r
+}
+
+// Next advances the reader to the next message.
+// Next returns false when the reader doesn't own any socket data
+// and a signal is received on doneReadingSignal,
+// or an error is encountered while reading.
+//
+// Callers must continue to call Next until it returns false.
+//
+// If the previous message was not fully read Next() panics.
+//
+// Next() panics if called on a reader created with SimpleReader().
+func (r *Reader) Next(doneReadingSignal chan struct{}) bool {
+	if r.toBeDeserialized == nil {
+		panic("called next on a simple reader")
+	}
+
+	if len(r.Buf) > 0 {
+		panic(fmt.Sprintf(
+			"cannot finish: unread data in buffer (message type: 0x%x)",
+			r.MsgType,
+		))
+	}
+
+	if r.data != nil && len(r.data.Buf) == 0 {
+		r.data.Release()
+		r.data = nil
+	}
+
+	r.MsgType = 0
+
+	if r.data == nil {
+		select {
+		case <-doneReadingSignal:
+			return false
+		case r.data = <-r.toBeDeserialized:
+			if r.data.Err != nil {
+				r.Err = r.data.Err
+				r.data.Release()
+				r.data = nil
+				return false
+			}
+		}
+	}
+
+	r.Err = r.feed(5)
+	if r.Err != nil {
+		return false
+	}
+
+	r.MsgType = r.PopUint8()
+	msgLen := int(r.PopUint32()) - 4
+
+	r.Err = r.feed(msgLen)
+	if r.Err != nil {
+		return false
+	}
+
+	(*reflect.SliceHeader)(unsafe.Pointer(&r.Buf)).Cap = msgLen
+	return true
+}
+
+func min(x, y int) int {
+	if x < y {
+		return x
+	}
+
+	return y
+}
+
+func (r *Reader) feed(n int) error {
+	if r.data != nil && len(r.data.Buf) == 0 {
+		r.data.Release()
+		r.data = nil
+	}
+
+	if n == 0 {
+		return nil
+	}
+
+	if r.data == nil {
+		r.data = <-r.toBeDeserialized
+
+		if r.data.Err != nil {
+			e := r.data.Err
+			r.data.Release()
+			r.data = nil
+			return e
+		}
+	}
+
+	m := min(n, len(r.data.Buf))
+	r.Buf = r.data.Buf[:m]
+	r.data.Buf = r.data.Buf[m:]
+
+	for len(r.Buf) < n {
+		previous := r.data
+		r.data = <-r.toBeDeserialized
+
+		if r.data.Err != nil {
+			previous.Release()
+			e := r.data.Err
+			r.data.Release()
+			r.data = nil
+			return e
+		}
+
+		m := min(n-len(r.Buf), len(r.data.Buf))
+		r.Buf = append(r.Buf, r.data.Buf[:m]...)
+		r.data.Buf = r.data.Buf[m:]
+		previous.Release()
+	}
+
+	return nil
+}
+
+// Discard skips n bytes.
+func (r *Reader) Discard(n int) {
+	r.Buf = r.Buf[n:]
+}
+
+// PopUint8 returns the next byte and advances the buffer.
+func (r *Reader) PopUint8() uint8 {
+	val := r.Buf[0]
+	r.Buf = r.Buf[1:]
+	return val
+}
+
+// PopUint16 reads a uint16 and advances the buffer.
+func (r *Reader) PopUint16() uint16 {
+	val := binary.BigEndian.Uint16(r.Buf[:2])
```

Comment:
Will `r.Buf[:2]` panic if there's only one byte in `r.Buf`?

Reply:
yes